File: chapter-03-sites/javascript/SiteOperations.js
/**
 * SiteOperations.js
 * Chapter 03: Sites
 *
 * SharePoint Site Operations Example
 * Demonstrates listing, creating, and retrieving SharePoint sites.
 *
 * Usa el módulo de auth compartido (common/graphAuth): el access token se inyecta
 * por constructor (DI), no se obtiene dentro de la clase.
 */

const axios = require('axios');
const { getAccessToken } = require('la-experiencia-sharepoint-code/graphAuth');

const GRAPH_BASE = 'https://graph.microsoft.com/v1.0';

class SiteOperations {
    /** @param {string} accessToken Bearer token for Microsoft Graph (inyectado). */
    constructor(accessToken) {
        if (!accessToken) throw new Error('Se requiere un access token para SiteOperations.');
        this.accessToken = accessToken;
    }

    /** @returns {Promise<{Authorization: string, 'Content-Type': string}>} */
    async getHeaders() {
        return {
            Authorization: `Bearer ${this.accessToken}`,
            'Content-Type': 'application/json',
        };
    }

    /** Lists all sites in the organization. */
    async listSites() {
        try {
            console.log('Fetching all sites...\n');
            const headers = await this.getHeaders();
            const response = await axios.get(`${GRAPH_BASE}/sites`, { headers });

            const sites = response.data.value;
            console.log(`Found ${sites.length} sites:`);
            console.log('-'.repeat(80));
            sites.forEach((site) => {
                console.log(`Display Name: ${site.displayName ?? site.name ?? 'N/A'}`);
                console.log(`  ID: ${site.id}`);
                console.log(`  Web URL: ${site.webUrl}`);
                console.log('-'.repeat(80));
            });
            return sites;
        } catch (error) {
            console.error('Error listing sites:', error.response?.data?.error?.message || error.message);
            throw error;
        }
    }

    /** Gets a specific site by hostname and site path (e.g. contoso.sharepoint.com / book-test). */
    async getSite(hostname, sitePath) {
        try {
            console.log(`Fetching site: ${hostname}/sites/${sitePath}`);
            const headers = await this.getHeaders();
            // Formato Graph path-based: /sites/{hostname}:/sites/{path}
            const url = `${GRAPH_BASE}/sites/${encodeURIComponent(hostname)}:/sites/${encodeURIComponent(sitePath)}`;
            const response = await axios.get(url, { headers });
            const site = response.data;

            console.log('\nSite found:');
            console.log(`  Display Name: ${site.displayName ?? site.name ?? 'N/A'}`);
            console.log(`  ID: ${site.id}`);
            console.log(`  Web URL: ${site.webUrl}`);
            console.log(`  Description: ${site.description ?? 'N/A'}`);
            return site;
        } catch (error) {
            console.error('Error getting site:', error.response?.data?.error?.message || error.message);
            throw error;
        }
    }

    /** Gets site by its unique identifier. */
    async getSiteById(siteId) {
        try {
            console.log(`Fetching site by ID: ${siteId}`);
            const headers = await this.getHeaders();
            const response = await axios.get(`${GRAPH_BASE}/sites/${siteId}`, { headers });
            const site = response.data;
            console.log('\nSite found:');
            console.log(`  Display Name: ${site.displayName ?? site.name ?? 'N/A'}`);
            console.log(`  Web URL: ${site.webUrl}`);
            return site;
        } catch (error) {
            console.error('Error getting site by ID:', error.response?.data?.error?.message || error.message);
            throw error;
        }
    }

    /** Gets the root site of the organization. */
    async getRootSite() {
        try {
            console.log('Fetching root site...');
            const headers = await this.getHeaders();
            const response = await axios.get(`${GRAPH_BASE}/sites/root`, { headers });
            const site = response.data;
            console.log('\nRoot site:');
            console.log(`  Display Name: ${site.displayName ?? site.name ?? 'N/A'}`);
            console.log(`  ID: ${site.id}`);
            console.log(`  Web URL: ${site.webUrl}`);
            return site;
        } catch (error) {
            console.error('Error getting root site:', error.response?.data?.error?.message || error.message);
            throw error;
        }
    }

    /** Searches for sites by keyword. */
    async searchSites(keyword) {
        try {
            console.log(`Searching for sites with keyword: '${keyword}'`);
            const headers = await this.getHeaders();
            const response = await axios.get(
                `${GRAPH_BASE}/sites?search=${encodeURIComponent(keyword)}`,
                { headers }
            );
            const sites = response.data.value;
            console.log(`\nFound ${sites.length} matching sites:`);
            console.log('-'.repeat(80));
            sites.forEach((site) => {
                console.log(`Display Name: ${site.displayName ?? site.name ?? 'N/A'}`);
                console.log(`  Web URL: ${site.webUrl}`);
                console.log(`  Description: ${site.description ?? 'N/A'}`);
                console.log('-'.repeat(80));
            });
            return sites;
        } catch (error) {
            console.error('Error searching sites:', error.response?.data?.error?.message || error.message);
            throw error;
        }
    }
}

/** Construye el token con el módulo común y ejecuta una demo de solo lectura contra book-test. */
async function main() {
    try {
        console.log('=== SharePoint Site Operations Example ===\n');
        const token = await getAccessToken();
        const siteOps = new SiteOperations(token);

        const hostname = process.env.SHAREPOINT_HOSTNAME || 'olddogsoft1.sharepoint.com';
        const sitePath = process.env.SHAREPOINT_SITE_PATH || 'book-test';

        await siteOps.getSite(hostname, sitePath);
        await siteOps.listSites();

        console.log('\nSite operations completed successfully!');
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

module.exports = { SiteOperations };

if (require.main === module) {
    main();
}

File: chapter-03-sites/python/site_operations.py
"""
site_operations.py
Chapter 03: Sites

SharePoint Site Operations Example
Demonstrates listing, creating, and retrieving SharePoint sites.

Usa el módulo de auth compartido (common/laexperiencia_sharepoint): el access token
se inyecta por constructor (DI), no se obtiene dentro de la clase.
"""

import sys
from typing import Dict, List, Optional

import requests

from laexperiencia_sharepoint import get_access_token

GRAPH_BASE = "https://graph.microsoft.com/v1.0"


class SiteOperations:
    """SharePoint Site Operations handler."""

    def __init__(self, access_token: str):
        if not access_token:
            raise ValueError("Se requiere un access token para SiteOperations.")
        self.access_token = access_token

    def _get_headers(self) -> Dict[str, str]:
        return {
            "Authorization": f"Bearer {self.access_token}",
            "Content-Type": "application/json",
        }

    def list_sites(self) -> List[Dict]:
        """Lists all sites in the organization."""
        try:
            print("Fetching all sites...\n")
            response = requests.get(f"{GRAPH_BASE}/sites", headers=self._get_headers())
            response.raise_for_status()

            sites = response.json()["value"]
            print(f"Found {len(sites)} sites:")
            print("-" * 80)
            for site in sites:
                print(f"Display Name: {site.get('displayName', site.get('name', 'N/A'))}")
                print(f"  ID: {site['id']}")
                print(f"  Web URL: {site['webUrl']}")
                print("-" * 80)
            return sites
        except requests.exceptions.RequestException as e:
            print(f"Error listing sites: {e}")
            raise

    def get_site(self, hostname: str, site_path: str) -> Optional[Dict]:
        """Gets a specific site by hostname and site path (e.g. contoso.sharepoint.com / book-test)."""
        try:
            print(f"Fetching site: {hostname}/sites/{site_path}")
            # Formato Graph path-based: /sites/{hostname}:/sites/{path}
            url = (
                f"{GRAPH_BASE}/sites/"
                f"{requests.utils.quote(hostname, safe='')}:/sites/"
                f"{requests.utils.quote(site_path, safe='')}"
            )
            response = requests.get(url, headers=self._get_headers())
            response.raise_for_status()

            site = response.json()
            print("\nSite found:")
            print(f"  Display Name: {site.get('displayName', site.get('name', 'N/A'))}")
            print(f"  ID: {site['id']}")
            print(f"  Web URL: {site['webUrl']}")
            print(f"  Description: {site.get('description', 'N/A')}")
            return site
        except requests.exceptions.RequestException as e:
            print(f"Error getting site: {e}")
            raise

    def get_site_by_id(self, site_id: str) -> Optional[Dict]:
        """Gets site by its unique identifier."""
        try:
            print(f"Fetching site by ID: {site_id}")
            response = requests.get(f"{GRAPH_BASE}/sites/{site_id}", headers=self._get_headers())
            response.raise_for_status()
            site = response.json()
            print("\nSite found:")
            print(f"  Display Name: {site.get('displayName', site.get('name', 'N/A'))}")
            print(f"  Web URL: {site['webUrl']}")
            return site
        except requests.exceptions.RequestException as e:
            print(f"Error getting site by ID: {e}")
            raise

    def get_root_site(self) -> Optional[Dict]:
        """Gets the root site of the organization."""
        try:
            print("Fetching root site...")
            response = requests.get(f"{GRAPH_BASE}/sites/root", headers=self._get_headers())
            response.raise_for_status()
            site = response.json()
            print("\nRoot site:")
            print(f"  Display Name: {site.get('displayName', site.get('name', 'N/A'))}")
            print(f"  ID: {site['id']}")
            print(f"  Web URL: {site['webUrl']}")
            return site
        except requests.exceptions.RequestException as e:
            print(f"Error getting root site: {e}")
            raise

    def search_sites(self, keyword: str) -> List[Dict]:
        """Searches for sites by keyword."""
        try:
            print(f"Searching for sites with keyword: '{keyword}'")
            response = requests.get(
                f"{GRAPH_BASE}/sites?search={keyword}", headers=self._get_headers()
            )
            response.raise_for_status()
            sites = response.json()["value"]
            print(f"\nFound {len(sites)} matching sites:")
            print("-" * 80)
            for site in sites:
                print(f"Display Name: {site.get('displayName', site.get('name', 'N/A'))}")
                print(f"  Web URL: {site['webUrl']}")
                print(f"  Description: {site.get('description', 'N/A')}")
                print("-" * 80)
            return sites
        except requests.exceptions.RequestException as e:
            print(f"Error searching sites: {e}")
            raise


def main():
    """Construye el token con el módulo común y ejecuta una demo de solo lectura contra book-test."""
    try:
        print("=== SharePoint Site Operations Example ===\n")
        import os

        token = get_access_token()
        site_ops = SiteOperations(token)

        hostname = os.environ.get("SHAREPOINT_HOSTNAME", "olddogsoft1.sharepoint.com")
        site_path = os.environ.get("SHAREPOINT_SITE_PATH", "book-test")

        site_ops.get_site(hostname, site_path)
        site_ops.list_sites()

        print("\nSite operations completed successfully!")
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()